src/igraph/io/objects.py

```diff
@@ -0,0 +1,813 @@
+from warnings import warn
+from collections import defaultdict
+
+
+from igraph.datatypes import UniqueIdGenerator
+from igraph.utils import first
+
+
+def _construct_graph_from_dict_list(
+    cls,
+    vertices,
+    edges,
+    directed=False,
+    vertex_name_attr="name",
+    edge_foreign_keys=("source", "target"),
+    iterative=False,
+):
+    """Constructs a graph from a list-of-dictionaries representation.
+
+    This representation assumes that vertices and edges are encoded in
+    two lists, each list containing a Python dict for each vertex and
+    each edge, respectively. A distinguished element of the vertex dicts
+    contain a vertex ID which is used in the edge dicts to refer to
+    source and target vertices. All the remaining elements of the dict
+    are considered vertex and edge attributes. Note that the implementation
+    does not assume that the objects passed to this method are indeed
+    lists of dicts, but they should be iterable and they should yield
+    objects that behave as dicts. So, for instance, a database query
+    result is likely to be fit as long as it's iterable and yields
+    dict-like objects with every iteration.
+
+    @param vertices: the data source for the vertices or C{None} if
+      there are no special attributes assigned to vertices and we
+      should simply use the edge list of dicts to infer vertex names.
+    @param edges: the data source for the edges.
+    @param directed: whether the constructed graph will be directed
+    @param vertex_name_attr: the name of the distinguished key in the
+      dicts in the vertex data source that contains the vertex names.
+      Ignored if C{vertices} is C{None}.
+    @param edge_foreign_keys: the name of the attributes in the dicts
+      in the edge data source that contain the source and target
+      vertex names.
+    @param iterative: whether to add the edges to the graph one by one,
```

Comment:
One could allow for an integer here that acts as a buffer, so you can:

```python
g = Graph.DictList(..., iterative=1000)
```
and it will call `add_edges` every 1000 edges (and once more at the end).

Dealing with all attributes correctly is a little tricky, so I opened a separate issue #436